http/server.go

```diff
@@ -13,304 +13,171 @@ package http
 	}
 }
 
-// Listen creates a new net.Listener and saves it on the receiver.
-func (s *Server) Listen(ctx context.Context) error {
-	var err error
-	if s.options.TLS.HasValue() {
-		return s.listenWithTLS(ctx)
+// WithTLSKeyPath sets the server TLS private key path.
+func WithTLSKeyPath(path string) ServerOpt {
+	return func(opts *ServerOptions) {
+		opts.TLSKeyPath = path
 	}
-
-	lc := net.ListenConfig{}
-	s.listener, err = lc.Listen(ctx, "tcp", s.Addr)
-	if err != nil {
-		return errors.WithStack(err)
-	}
-
-	// Save the address on the server in case the port was set to random
-	// and that we want to see what was assigned.
-	s.address = s.listener.Addr().String()
-
-	return nil
 }
 
-func (s *Server) listenWithTLS(ctx context.Context) error {
-	cfg := &tls.Config{
-		MinVersion: tls.VersionTLS12,
-		// We only allow cipher suites that are marked secure
-		// by ssllabs
-		CipherSuites: []uint16{
-			tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
-			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
-			tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
-			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
-		},
-		ServerName: "DefraDB",
-	}
-
-	if s.options.Domain.HasValue() && s.options.Domain.Value() != "" {
-		s.Addr = s.options.TLS.Value().Port
-
-		if s.options.TLS.Value().Email == "" || s.options.TLS.Value().Email == config.DefaultAPIEmail {
-			return ErrNoEmail
-		}
-
-		certCache := path.Join(s.options.RootDir, "autocerts")
-
-		log.FeedbackInfo(
-			ctx,
-			"Generating auto certificate",
-			logging.NewKV("Domain", s.options.Domain.Value()),
-			logging.NewKV("Certificate cache", certCache),
-		)
-
-		m := &autocert.Manager{
-			Cache:      autocert.DirCache(certCache),
-			Prompt:     autocert.AcceptTOS,
-			Email:      s.options.TLS.Value().Email,
-			HostPolicy: autocert.HostWhitelist(s.options.Domain.Value()),
-		}
-
-		cfg.GetCertificate = m.GetCertificate
-
-		// We set manager on the server instance to later start
-		// a redirection server.
-		s.certManager = m
-	} else {
-		// When not using auto cert, we create a self signed certificate
-		// with the provided public and prive keys.
-		log.FeedbackInfo(ctx, "Generating self signed certificate")
-
-		cert, err := tls.LoadX509KeyPair(
-			s.options.TLS.Value().PrivateKey,
-			s.options.TLS.Value().PublicKey,
-		)
-		if err != nil {
-			return NewErrFailedToLoadKeys(err, s.options.TLS.Value().PublicKey, s.options.TLS.Value().PrivateKey)
-		}
-
-		cfg.Certificates = []tls.Certificate{cert}
-	}
-
-	var err error
-	s.listener, err = tls.Listen("tcp", s.Addr, cfg)
-	if err != nil {
-		return errors.WithStack(err)
-	}
-
-	// Save the address on the server in case the port was set to random
-	// and that we want to see what was assigned.
-	s.address = s.listener.Addr().String()
-
-	return nil
+// Server struct holds the Handler for the HTTP API.
+type Server struct {
+	address atomic.Value
```

Comment:
question: Why is this atomic? It only appears to be set once, in `ListenAndServe`, and that function cannot be called multiple times concurrently no?

todo: If it remains atomic, please document why.

Reply:
I've added documentation to clarify the use of the atomic value.

Reply:
Ah, you want `AssignedAddr` to return empty if `ListenAndServe` is in progress?  Would it not be better to lock, and block `AssignedAddr` from returning until the address has been assigned? (i.e. no atomic, but a RWLock)

Reply:
I agree that would be better, but I realized we don't use this method anywhere outside of `server_test.go`. I've removed the function. 

As a side note: tests that need a random port assignment should use the `httptest` package instead.

Reply:
> As a side note: tests that need a random port assignment should use the httptest package instead.

That's a really nice change. Thanks for removing the need to keep track of the randomly assigned port on the http server :)